sklearn/mixture/base.py

```diff
@@ -0,0 +1,425 @@
+
+
+def _check_X(X, n_components=None, n_features=None):
+    """Check the input data X.
+
+    Parameters
+    ----------
+    X : array-like, (n_samples, n_features)
+
+    n_components : int
+
+    Returns
+    -------
+    X : array, (n_samples, n_features)
+    """
+    X = check_array(X, dtype=[np.float64, np.float32])
+    if n_components is not None and X.shape[0] < n_components:
+        raise ValueError('Expected n_samples >= n_components'
+                         'but got n_components = %d, n_samples = %d'
+                         % (n_components, X.shape[0]))
+    if n_features is not None and X.shape[1] != n_features:
+        raise ValueError("Expected the input data X have %d features, "
+                         "but got %d features"
+                         % (n_features, X.shape[1]))
+    return X
+
+
+class MixtureBase(six.with_metaclass(ABCMeta, DensityMixin, BaseEstimator)):
+    """Base class for mixture models.
+
+    This abstract class specifies an interface for all mixture classes and
+    provides basic common methods for mixture models.
+    """
+
+    def __init__(self, n_components, tol, reg_covar,
+                 max_iter, n_init, init_params, random_state, warm_start,
+                 verbose, verbose_interval):
+        self.n_components = n_components
+        self.tol = tol
+        self.reg_covar = reg_covar
+        self.random_state = random_state
+        self.max_iter = max_iter
+        self.n_init = n_init
+        self.init_params = init_params
+        self.warm_start = warm_start
+        self.verbose = verbose
+        self.verbose_interval = verbose_interval
+
+        self.n_features = None
+        self.converged_ = False
+        self.n_iter_ = 0
+        self.initialized_ = False
+
+    def _check_initial_parameters(self):
+        """Check values of the basic parameters."""
+        if self.n_init < 1:
+            raise ValueError("Invalid value for 'n_init': %d "
+                             "Estimation requires at least one run"
+                             % self.n_init)
+
+        if self.max_iter < 1:
+            raise ValueError("Invalid value for 'max_iter': %d "
+                             "Estimation requires at least one iteration"
+                             % self.max_iter)
+
+        if self.reg_covar < 0:
+            raise ValueError("Invalid value for 'reg_covar': %.5f "
+                             "regularization on covariance must be "
+                             "non-negative"
+                             % self.reg_covar)
+
+        # Check all the parameters values of the derived class
+        self._check_parameters()
+
+    @abstractmethod
+    def _check_parameters(self):
+        """Check initial parameters of the derived class."""
+        pass
+
+    def _initialize_parameters(self, X):
+        """Initialize the model parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
```

Comment:
I know we're not always consistent but remove the =

shape (n_samples, n_features)